File: saksflyt/src/test/java/no/nav/melosys/saksflyt/impl/ProsessinstansOpprettetListenerTest.java
package no.nav.melosys.saksflyt.impl;

import java.util.Set;
import java.util.UUID;

import no.nav.melosys.domain.saksflyt.ProsessStatus;
import no.nav.melosys.domain.saksflyt.Prosessinstans;
import no.nav.melosys.domain.saksflyt.ProsessinstansInfo;
import no.nav.melosys.domain.saksflyt.ProsessinstansLåsType;
import no.nav.melosys.repository.ProsessinstansRepository;
import no.nav.melosys.saksflyt.api.ProsessinstansBehandler;
import no.nav.melosys.service.saksflyt.ProsessinstansOpprettetEvent;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ProsessinstansOpprettetListenerTest {

    @Mock
    private ProsessinstansRepository prosessinstansRepository;
    @Mock
    private ProsessinstansBehandler prosessinstansBehandler;

    private ProsessinstansOpprettetListener prosessinstansOpprettetListener;

    private Prosessinstans prosessinstans;

    @BeforeEach
    void setup() {
        prosessinstansOpprettetListener = new ProsessinstansOpprettetListener(prosessinstansBehandler, prosessinstansRepository);
        prosessinstans = new Prosessinstans();
        prosessinstans.setId(UUID.randomUUID());
    }

    @Test
    void oppdaterProsessinstansstatus_harIkkeLås_settesIkkePåVent() {
        prosessinstans.setStatus(ProsessStatus.KLAR);
        prosessinstansOpprettetListener.oppdaterProsessinstansstatus(new ProsessinstansOpprettetEvent(prosessinstans));
        assertThat(prosessinstans.getStatus()).isEqualTo(ProsessStatus.KLAR);
        verifyNoInteractions(prosessinstansRepository, prosessinstansBehandler);
    }

    @Test
    void oppdaterProsessinstansstatus_finnesProsessMedSammeReferanseUnderBehandling_settesIkkePåVent() {
        prosessinstans.setStatus(ProsessStatus.KLAR);
        final var låsReferanse = "12_12_1";
        prosessinstans.setLåsType(ProsessinstansLåsType.SED);
        prosessinstans.setLåsReferanse(låsReferanse);

        var eksisterendeProsessinstans = prosessinstans(låsReferanse, ProsessStatus.UNDER_BEHANDLING);
        when(prosessinstansRepository.findAllByStatusNotInAndLåsReferanseStartingWith(any(), any())).thenReturn(Set.of(new ProsessinstansInfo(eksisterendeProsessinstans)));

        prosessinstansOpprettetListener.oppdaterProsessinstansstatus(new ProsessinstansOpprettetEvent(prosessinstans));
        assertThat(prosessinstans.getStatus()).isEqualTo(ProsessStatus.KLAR);
    }

    @Test
    void oppdaterProsessinstansstatus_finnesProsessMedSammeReferanseUlikId_settesPåVent() {
        prosessinstans.setStatus(ProsessStatus.KLAR);
        final var låsReferanse = "12_12_1";
        prosessinstans.setLåsType(ProsessinstansLåsType.SED);
        prosessinstans.setLåsReferanse(låsReferanse);

        var eksisterendeProsessinstans = prosessinstans("12_13_1", ProsessStatus.UNDER_BEHANDLING);
        when(prosessinstansRepository.findAllByStatusNotInAndLåsReferanseStartingWith(any(), any())).thenReturn(Set.of(new ProsessinstansInfo(eksisterendeProsessinstans)));

        prosessinstansOpprettetListener.oppdaterProsessinstansstatus(new ProsessinstansOpprettetEvent(prosessinstans));
        assertThat(prosessinstans.getStatus()).isEqualTo(ProsessStatus.PÅ_VENT);
    }

    @Test
    void behandleOpprettetProsessinstans_statusErKlar_behandlesVidere() {
        prosessinstans.setStatus(ProsessStatus.KLAR);
        prosessinstansOpprettetListener.behandleOpprettetProsessinstans(new ProsessinstansOpprettetEvent(prosessinstans));
        verify(prosessinstansBehandler).behandleProsessinstans(prosessinstans);
    }

    @Test
    void behandleOpprettetProsessinstans_statusErPåVent_behandlesIkke() {
        prosessinstans.setStatus(ProsessStatus.PÅ_VENT);
        prosessinstansOpprettetListener.behandleOpprettetProsessinstans(new ProsessinstansOpprettetEvent(prosessinstans));
        verifyNoInteractions(prosessinstansBehandler);
    }

    private Prosessinstans prosessinstans(String låsReferanse, ProsessStatus prosessStatus) {
        var prosessinstans = new Prosessinstans();
        prosessinstans.setId(UUID.randomUUID());
        prosessinstans.setLåsType(ProsessinstansLåsType.SED);
        prosessinstans.setLåsReferanse(låsReferanse);
        prosessinstans.setStatus(prosessStatus);
        return prosessinstans;
    }
}

File: saksflyt/src/test/kotlin/no/nav/melosys/saksflyt/ProsessinstansOpprettetListenerTest.kt
package no.nav.melosys.saksflyt

import io.mockk.Runs
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.slot
import io.mockk.verify
import no.nav.melosys.saksflytapi.ProsessinstansOpprettetEvent
import no.nav.melosys.saksflytapi.domain.ProsessStatus
import no.nav.melosys.saksflytapi.domain.ProsessType
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@ExtendWith(MockKExtension::class)
class ProsessinstansOpprettetListenerTest {

    @MockK
    lateinit var prosessinstansBehandlerDelegate: ProsessinstansBehandlerDelegate

    @MockK
    lateinit var prosessinstansBehandler: ProsessinstansBehandler

    @MockK
    lateinit var prosessinstansRepository: ProsessinstansRepository

    @InjectMockKs
    lateinit var listener: ProsessinstansOpprettetListener

    private lateinit var prosessinstansId: UUID
    private lateinit var eventProsessinstans: Prosessinstans
    private lateinit var freshProsessinstans: Prosessinstans
    private lateinit var event: ProsessinstansOpprettetEvent

    @BeforeEach
    fun setUp() {
        prosessinstansId = UUID.randomUUID()

        // Prosessinstans fra eventen (stale referanse fra HTTP-transaksjon)
        // Simulerer at denne ble opprettet tidligere og har gammel endretDato
        eventProsessinstans = Prosessinstans.builder()
            .medId(prosessinstansId)
            .medType(ProsessType.IVERKSETT_VEDTAK_FTRL)
            .medStatus(ProsessStatus.KLAR)
            .medEndretDato(java.time.LocalDateTime.of(2024, 1, 1, 10, 0))
            .build()

        // Fersk prosessinstans fra databasen - har nyere endretDato
        // Dette simulerer at entiteten har blitt oppdatert mellom event-publisering og AFTER_COMMIT
        freshProsessinstans = Prosessinstans.builder()
            .medId(prosessinstansId)
            .medType(ProsessType.IVERKSETT_VEDTAK_FTRL)
            .medStatus(ProsessStatus.KLAR)
            .medEndretDato(java.time.LocalDateTime.of(2024, 1, 1, 12, 0))
            .build()

        event = ProsessinstansOpprettetEvent(eventProsessinstans)
    }

    @Test
    fun `behandleOpprettetProsessinstans laster prosessinstans fra database`() {
        // Given
        every { prosessinstansRepository.findById(prosessinstansId) } returns Optional.of(freshProsessinstans)
        every { prosessinstansBehandler.behandleProsessinstans(any()) } just Runs

        // When
        listener.behandleOpprettetProsessinstans(event)

        // Then - verifiser at findById ble kalt med riktig ID
        verify { prosessinstansRepository.findById(prosessinstansId) }

        // Verifiser at behandleProsessinstans ble kalt med prosessinstansen fra repository (ikke fra event)
        val capturedProsessinstans = slot<Prosessinstans>()
        verify { prosessinstansBehandler.behandleProsessinstans(capture(capturedProsessinstans)) }

        // Den fangede prosessinstansen skal være den fra repository, ikke den fra event
        // Vi sjekker dette ved å verifisere at det er samme referanse som freshProsessinstans
        assert(capturedProsessinstans.captured === freshProsessinstans) {
            "Forventet at prosessinstans fra repository ble brukt, ikke fra event"
        }

        // Ekstra verifisering: sjekk at det IKKE er event-objektet
        assert(capturedProsessinstans.captured !== eventProsessinstans) {
            "Prosessinstans fra event skal IKKE brukes - den kan ha stale data"
        }

        // Verifiser at vi faktisk brukte objektet med nyere endretDato (fra DB)
        assert(capturedProsessinstans.captured.endretDato == freshProsessinstans.endretDato) {
            "Forventet at prosessinstans med oppdatert endretDato ble brukt"
        }
    }

    @Test
    fun `behandleOpprettetProsessinstans starter ikke behandling når prosessinstans er på vent`() {
        // Given
        val prosessinstansPåVent = Prosessinstans.builder()
            .medId(prosessinstansId)
            .medType(ProsessType.IVERKSETT_VEDTAK_FTRL)
            .medStatus(ProsessStatus.PÅ_VENT)
            .build()

        every { prosessinstansRepository.findById(prosessinstansId) } returns Optional.of(prosessinstansPåVent)

        // When
        listener.behandleOpprettetProsessinstans(event)

        // Then
        verify { prosessinstansRepository.findById(prosessinstansId) }
        verify(exactly = 0) { prosessinstansBehandler.behandleProsessinstans(any()) }
    }

    @Test
    fun `behandleOpprettetProsessinstans kaster exception når prosessinstans ikke finnes`() {
        // Given
        every { prosessinstansRepository.findById(prosessinstansId) } returns Optional.empty()

        // When/Then
        assertThrows<IllegalStateException> {
            listener.behandleOpprettetProsessinstans(event)
        }

        verify(exactly = 0) { prosessinstansBehandler.behandleProsessinstans(any()) }
    }

    @Test
    fun `oppdaterProsessinstansstatus bruker prosessinstans fra event`() {
        // Given - BEFORE_COMMIT bruker fortsatt event-objektet siden det er i samme transaksjon
        every { prosessinstansBehandlerDelegate.oppdaterStatusOmSkalPåVent(any()) } just Runs

        // When
        listener.oppdaterProsessinstansstatus(event)

        // Then
        verify { prosessinstansBehandlerDelegate.oppdaterStatusOmSkalPåVent(eventProsessinstans) }
    }
}
